packages/xo-server/src/xapi-stats.js

```diff
@@ -411,40 +202,57 @@ export default class XapiStats {
       .then(response => response.readAll().then(JSON5.parse))
   }
 
-  async _getLastTimestamp (xapi, host, step) {
-    if (this._hosts[host.address][step] === undefined) {
-      const serverTimeStamp = await getServerTimestamp(xapi, host)
-      return serverTimeStamp - step * RRD_POINTS_PER_STEP[step] + step
-    }
+  async _getNextTimestamp (xapi, host, step) {
+    const currentTimeStamp = await getServerTimestamp(xapi, host.$ref)
+    const maxDuration = step * RRD_POINTS_PER_STEP[step]
+    const lastTimestamp = get(this._statsByObject, [
+      host.uuid,
+      step,
+      'endTimestamp',
+    ])
 
-    return this._hosts[host.address][step].endTimestamp
+    if (
+      lastTimestamp === undefined ||
+      currentTimeStamp - lastTimestamp + step > maxDuration
+    ) {
+      return currentTimeStamp - maxDuration + step
+    }
+    return lastTimestamp
   }
 
-  _getPoints (hostname, step, vmId) {
-    const hostStats = this._hosts[hostname][step]
+  _getStats (hostUuid, step, vmUuid) {
+    const hostStats = this._statsByObject[hostUuid][step]
 
-    // Return host points
-    if (vmId === undefined) {
+    // Return host stats
+    if (vmUuid === undefined) {
       return {
         interval: step,
         ...hostStats,
       }
     }
 
-    const vmsStats = this._vms[hostname][step]
-
-    // Return vm points
+    // Return vm stats
     return {
       interval: step,
       endTimestamp: hostStats.endTimestamp,
-      stats: (vmsStats && vmsStats[vmId]) || getNewVmStats(),
+      ...this._statsByObject[vmUuid][step],
+    }
```

Comment:
Why are you getting `hostStats` in this branch?

Reply:
`hostStats`  is available in this branch because we query the RRD by using the host's address.
I thought than is better to get the `endTimestamp` from the hostStats than doing a loop to assign it for each vm.